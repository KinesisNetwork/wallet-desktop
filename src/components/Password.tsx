import { Wallet } from '@types'
import * as ClipboardJS from 'clipboard'
import * as React from 'react'

export interface Props {
  isAccountUnlocked: boolean
  activeWallet: Wallet
  password: string
  decryptedPrivateKey: string
  unlockWallet: (wallet: Wallet, password: string) => any
  setPasswordInput: (input: string) => any
  lockWallet: (wallet: Wallet) => any
}

const LockedWallet: React.SFC<Props> = ({ password, unlockWallet, activeWallet, setPasswordInput }) => (
  <div>
    <div className='field has-addons has-addons-centered'>
      <div className='control has-icons-left'>
        <input
          id='wallet-unlock-password'
          className='input'
          value={password}
          type='password'
          placeholder='Unlock Account'
          onChange={(e) => setPasswordInput(e.currentTarget.value)}
        />
        <span className='icon is-left has-text-grey-lighter'><i className='fas fa-key' /></span>
      </div>
      <div className='control'>
        <button type='button' className='button' onClick={() => unlockWallet(activeWallet, password)}>
          <span className='icon'>
            <i className='fas fa-lock-open' />
          </span>
        </button>
      </div>
    </div>
  </div>
)

class UnlockedWallet extends React.Component<Props, Readonly<{}>> {
  public copyBtn
  public clipboard

  public componentDidMount() {
    this.clipboard = new ClipboardJS(this.copyBtn)
  }

  public componentWillUnmount() {
    this.clipboard.destroy()
  }

  public render() {
    return (
      <div>
        <div className='field is-grouped is-grouped-centered'>
          <div className='control'>
            <button
              ref={(ref) => this.copyBtn = ref}
              type='button'
              className='button'
              data-clipboard-text={this.props.decryptedPrivateKey}
            >
              <span className='icon'>
                <i className='fas fa-copy' />
              </span>
              <span>Private Key</span>
            </button>
          </div>
          <div className='control'>
            <button type='submit' className='button' onClick={() => this.props.lockWallet(this.props.activeWallet)}>
              <span className='icon'>
                <i className='fas fa-lock' />
              </span>
            </button>
          </div>
        </div>
      </div>
    )
  }
}

export const Password: React.SFC<Props> = (props) => (
  <React.Fragment>
    {props.isAccountUnlocked ? <UnlockedWallet {...props} /> : <LockedWallet {...props} />}
  </React.Fragment>
)

Password.displayName = 'Password'
