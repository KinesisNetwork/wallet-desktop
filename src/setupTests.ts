import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
