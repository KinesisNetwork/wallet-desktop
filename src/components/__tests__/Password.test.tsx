import { shallow } from 'enzyme';
import * as React from 'react'
import * as sinon from 'sinon'

import { LockedWallet, Password, UnlockedWallet } from '@components/Password'
import '../../setupTests'

describe('Password', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Password
        activeWallet={
          {
            accountName: '',
            encryptedPrivateKey: '',
            publicKey: ''
          }
        }
        decryptedPrivateKey=''
        isAccountUnlocked={false}
        password=''
        unlockWallet={() => null}
        setPasswordInput={() => null}
        lockWallet={() => null}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render LockedWallet if isAccountUnlocked is false', () => {
    const wrapper = shallow(
      <Password
        activeWallet={
          {
            accountName: '',
            encryptedPrivateKey: '',
            publicKey: ''
          }
        }
        decryptedPrivateKey=''
        isAccountUnlocked={false}
        password=''
        unlockWallet={() => null}
        setPasswordInput={() => null}
        lockWallet={() => null}
      />
    )

    expect(wrapper.find('LockedWallet')).toHaveLength(1)
  })

  it('renders UnlockedWallet component if isAccountUnlocked is true', () => {
    const wrapper = shallow(
      <Password
        activeWallet={
          {
            accountName: '',
            encryptedPrivateKey: '',
            publicKey: ''
          }
        }
        decryptedPrivateKey=''
        isAccountUnlocked={true}
        password=''
        unlockWallet={() => null}
        setPasswordInput={() => null}
        lockWallet={() => null}
      />
    )

    expect(wrapper.find('UnlockedWallet')).toHaveLength(1)
  })
})

describe('UnlockWallet', () => {
  let componentDidMountStub
  let props

  beforeEach(() => {
    props = {
      activeWallet: {
        accountName: '',
        encryptedPrivateKey: '',
        publicKey: ''
      },
      decryptedPrivateKey: 'decrypted',
      isAccountUnlocked: true,
      password: '',
      unlockWallet: () => null,
      setPasswordInput: () => null,
      lockWallet: () => null,
    }
    componentDidMountStub = sinon.stub(UnlockedWallet.prototype, 'componentDidMount')
    componentDidMountStub.callsFake(() => null)
  })

  afterEach(() => {
    componentDidMountStub.restore()
  })

  it('renders correctly', () => {
    const wrapper = shallow(<UnlockedWallet {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders two buttons', () => {
    const wrapper = shallow(<UnlockedWallet {...props} />)

    const buttons = wrapper.find('button')

    expect(buttons).toHaveLength(2)
    expect(buttons.at(0).prop('data-clipboard-text')).toEqual('decrypted')
    expect(buttons.at(0).text()).toEqual('Private Key')
    expect(buttons.at(1).find('i').hasClass('fa-lock')).toBe(true)
  })

  it('calls the lockWallet method on clicking the second button', () => {
    const lockWalletMock = jest.fn()
    const lockWalletMockArgument = {
      accountName: '',
      encryptedPrivateKey: '',
      publicKey: ''
    }
    const wrapper = shallow(<UnlockedWallet {...props} lockWallet={lockWalletMock} />)

    const button = wrapper.find('button').at(1)
    button.simulate('click')

    expect(lockWalletMock).toHaveBeenCalledWith(lockWalletMockArgument)
  })
})

describe('LockedWallet', () => {
  let props

  beforeEach(() => {
    props = {
      activeWallet: {
        accountName: '',
        encryptedPrivateKey: '',
        publicKey: ''
      },
      decryptedPrivateKey: '',
      isAccountUnlocked: false,
      password: '123',
      unlockWallet: () => null,
      setPasswordInput: () => null,
      lockWallet: () => null,
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(<LockedWallet {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should render an input field and an unlock button', () => {
    const wrapper = shallow(<LockedWallet {...props} />)

    const input = wrapper.find('input')
    const button = wrapper.find('button')

    expect(input).toHaveLength(1)
    expect(input.prop('value')).toEqual('123')
    expect(button).toHaveLength(1)
    expect(button.find('i')).toHaveLength(1)
    expect(button.find('i').hasClass('fa-lock-open')).toBe(true)
  })

  it('triggers setPasswordInput method on change event', () => {
    const setPasswordInputMock = jest.fn()
    const wrapper = shallow(<LockedWallet {...props} setPasswordInput={setPasswordInputMock} />)

    const input = wrapper.find('input')
    input.simulate('change', { currentTarget: { value: '123' } })

    expect(setPasswordInputMock).toHaveBeenCalledWith('123')
  })

  it('triggers unLockWallet method on clicking the button', () => {
    const unLockWalletMock = jest.fn()
    const unLockWalletMockArgument = {
      accountName: '',
      encryptedPrivateKey: '',
      publicKey: ''
    }
    const wrapper = shallow(<LockedWallet {...props} unlockWallet={unLockWalletMock} />)

    const button = wrapper.find('button')
    button.simulate('click')
    expect(unLockWalletMock).toHaveBeenCalledWith(unLockWalletMockArgument, '123')
  })
})
