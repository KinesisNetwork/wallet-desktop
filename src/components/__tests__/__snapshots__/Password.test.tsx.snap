// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LockedWallet renders correctly 1`] = `
<div>
  <div
    className="field has-addons has-addons-centered"
  >
    <div
      className="control has-icons-left"
    >
      <input
        className="input"
        id="wallet-unlock-password"
        onChange={[Function]}
        placeholder="Unlock Account"
        type="password"
        value="123"
      />
      <span
        className="icon is-left has-text-grey-lighter"
      >
        <i
          className="fas fa-key"
        />
      </span>
    </div>
    <div
      className="control"
    >
      <button
        className="button"
        onClick={[Function]}
        type="button"
      >
        <span
          className="icon"
        >
          <i
            className="fas fa-lock-open"
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Password renders correctly 1`] = `
<Fragment>
  <LockedWallet
    activeWallet={
      Object {
        "accountName": "",
        "encryptedPrivateKey": "",
        "publicKey": "",
      }
    }
    decryptedPrivateKey=""
    isAccountUnlocked={false}
    lockWallet={[Function]}
    password=""
    setPasswordInput={[Function]}
    unlockWallet={[Function]}
  />
</Fragment>
`;

exports[`UnlockWallet renders correctly 1`] = `
<div>
  <div
    className="field is-grouped is-grouped-centered"
  >
    <div
      className="control"
    >
      <button
        className="button"
        data-clipboard-text="decrypted"
        type="button"
      >
        <span
          className="icon"
        >
          <i
            className="fas fa-copy"
          />
        </span>
        <span>
          Private Key
        </span>
      </button>
    </div>
    <div
      className="control"
    >
      <button
        className="button"
        onClick={[Function]}
        type="submit"
      >
        <span
          className="icon"
        >
          <i
            className="fas fa-lock"
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;
